tsdb/engine/tsm1/cache.go

```diff
@@ -180,37 +187,55 @@ func (c *Cache) Keys() []string {
 
 // Values returns a copy of all values, deduped and sorted, for the given key.
 func (c *Cache) Values(key string) Values {
-	values, needSort := func() (Values, bool) {
-		c.mu.RLock()
-		defer c.mu.RUnlock()
-		e := c.store[key]
-		if e == nil {
-			return nil, false
+	c.mu.Lock()
```

Comment:
Exactly. If you can do it with a read lock (i.e. no sorting needed), do so, otherwise use a write lock.